### PR TITLE
adding probabilistic BKZ simulator from BSW18

### PR DESCRIPTION
**Pull request by fvirdia**
The simulator has the same interface as the CN11 simulator already present, plus a new function for outputting the average over many runs. The code is based on the paper description. The code for the simulator and from comparing it to the output from the simulator as originally implemented can be found in [https://github.com/fvirdia/BSW18-simulator](https://github.com/fvirdia/BSW18-simulator).

### CODE REVIEW

File: src/fpylll/tools/probabilistic_bkz_simulator.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
+# -*- coding: utf-8 -*-
+"""
+BKZ simulation algorithm as proposed in
+
+- Bai, S., & Stehlé, D. & Wen, W. (2018). Measuring, simulating and exploiting
+  the head concavity phenomenon in BKZ. In T. Peyrin, & S. Galbraith,
+  ASIACRYPT~2018 : Springer, Heidelberg.
+
+The code is based on the implementation of the simulation algorithm proposed by
```
Comment (malb) [inline src/fpylll/tools/probabilistic_bkz_simulator.py:9]: Should this be here?

File: src/fpylll/tools/probabilistic_bkz_simulator.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+
+from fpylll.tools.quality import basis_quality
+from fpylll.tools.bkz_stats import pretty_dict
+from fpylll.fplll.bkz import BKZ
+from fpylll.fplll.integer_matrix import IntegerMatrix
+from fpylll.fplll.gso import MatGSO, GSO
+from fpylll import FPLLL
+
+
+rk = (
```
Comment (malb) [inline src/fpylll/tools/probabilistic_bkz_simulator.py:30]: I'd suggest to have this list only once in the library.

File: src/fpylll/tools/probabilistic_bkz_simulator.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+            r["i"] = j
+            for k, v in basis_quality(list(map(lambda x: 2.0 ** (2 * x), l))).items():
+                r[k] = v
+            print(pretty_dict(r))
+
+    l = list(map(lambda x: 2.0 ** (2 * x), l))
+    return l, j + 1
+
+
+def averaged_simulate(L, params_fplll, tries=10):
```
Comment (malb) [inline src/fpylll/tools/probabilistic_bkz_simulator.py:192]: We're a bit punkrock in general with `param` v `params` but I suggest to unify it in this file. I tend towards `params`

**Comment by malb** [comment]
Can you give some indication on how it differs?
- running time
- quality of output (IIRC the difference only really kicks in after many tours?)

File: src/fpylll/tools/probabilistic_bkz_simulator.py (hunk 1 above)
Comment (fvirdia) [reply src/fpylll/tools/probabilistic_bkz_simulator.py:9]: What does "this" refer to?

File: src/fpylll/tools/probabilistic_bkz_simulator.py (hunk 2 above)
Comment (fvirdia) [reply src/fpylll/tools/probabilistic_bkz_simulator.py:30]: will change

File: src/fpylll/tools/probabilistic_bkz_simulator.py (hunk 3 above)
Comment (fvirdia) [reply src/fpylll/tools/probabilistic_bkz_simulator.py:192]: sure

File: src/fpylll/tools/probabilistic_bkz_simulator.py (hunk 1 above)
Comment (malb) [reply src/fpylll/tools/probabilistic_bkz_simulator.py:9]: The "The code is based on" comment, it's based o a different simulator, no?

Comment (fvirdia) [reply src/fpylll/tools/probabilistic_bkz_simulator.py:9]: By that comment I mean that I opened the CN11 simulator source in fpylll, and made the changes following the paper. Then I compared to the original implementation that Shi pointed me to after I had written their simulator.

**Comment by shi-bai** [comment]
Hi Fernando, Martin,

As discussed, could you please check the following (substitute) PR? It fixes the determinant problem as well as make it closer to actual case for smaller blocksizes. The two issues we discussed is the funky plot (I think that's some issue not on the indices -- but some normalization of the det for the last block). The second issue is regarding the behavior of the simulation when the blocksize is small (say < 20). The determinant now is invariant during the simulation and the practical behavior of the simulation looks similar to CN but does show the head concavity. It does not have the funky plot in my tests. In either case, the behavior of the simulator for larger blocksizes matches the CN simulator which is perhaps what we really concerns with.

 Could you please let me know if that works with some of your test? I provide some experiments here.

[Uploading bkz_simulator.py.txt…]()

**Comment by shi-bai** [comment]
Here are some plots, fixing n = 100 and various beta = 99, 90, 60, 45, 30, 20, 10, 4, number of tours = beta. I compared that to CN simulator. For small blocksizes, there are some variation but that matches theory I guess..

**Comment by fvirdia** [comment]
Hey, I tried the the new code, the outputs look the same as the older one at the naked eye, for random params, but now with constant volume, which is cool. Timing seems good, I'll report on that later martin.

I think the plots that Shi reports as "averaged over 10" are actually averaged over 1 (`tries=1` in the test code). I changed my mind on how to average the outputs. Currently, the code averages the squared norms, but I think maybe we should average the logs, this way the volume would stay constant.

**Comment by shi-bai** [comment]
> Hey, I tried the the new code, the outputs look the same as the older one at the naked eye, for random params, but now with constant volume, which is cool. Timing seems good, I'll report on that later.
> 
> I think the plots that Shi reports as "averaged over 10" are actually averaged over 1 (`tries=1` in the test code). I changed my mind on how to average the outputs. Currently, the code averages the squared norms, but I think maybe we should average the logs, this way the volume would stay constant.

Fernando, thanks for experiments.

For the tries=1, I remembered I used 10 in experiments (the plots is somewhat averaged), perhaps i hard-coded 10 in the code somewhere..

**Comment by malb** [comment]
We're heading for a review, is your ambition to get this in for that?

**Comment by fvirdia** [comment]
I think we could. The plan would be merging the two simultors inside `bkz_simulator.py` as Shi did.

I have made the change locally, and timed the simulators, verifying that CN11 stays the same.
The results are kinda funny, not sure if it's just python caching something, or something inherent to the simulator. You can see below the results, I created an q-ary basis of dimension `n`, rank `n/2`, `q = 3329` and simulated `tau` tours of BKZ. I report values in seconds, as reported by the `timeit` module, averaged over 20 tries. `cn11 old` is the upstream code, `cn11 new` is the code after merging the modules, the outputs are  checked to be identical for these two.

```
n       bs      tau     CN11 old        CN11 new        BSW18           BSW18 (avg)     BSW/CN
100     4       4       0.000821        0.000814        0.001825        0.017071        2.242778
100     4       20      0.003627        0.003633        0.007323        0.027452        2.015370
100     4       2000    0.350724        0.351395        0.007248        0.027447        0.020627
100     15      20      0.003287        0.003294        0.012874        0.128750        3.908746
100     15      2000    0.055281        0.048973        0.073710        0.557609        1.505112
100     20      20      0.003124        0.003147        0.014248        0.143857        4.527200
100     30      20      0.002842        0.002862        0.016554        0.160904        5.783534
100     40      20      0.002552        0.002563        0.016805        0.166156        6.557121
100     50      20      0.002338        0.002337        0.017729        0.180814        7.587125
100     60      20      0.002422        0.002421        0.019791        0.198907        8.175263
100     70      20      0.001906        0.001882        0.021628        0.215455        11.493471
100     80      20      0.001946        0.001953        0.022331        0.227711        11.432949
100     90      20      0.001306        0.001270        0.023225        0.235361        18.290566
```

**Comment by malb** [comment]
The plan is to release tomorrow, do you think the PR will be ready by then?

**Comment by fvirdia** [comment]
Have pushed the changes and am waiting for checks to pass/making changes as necessary.

**Pull request merged**